tests/tls_handler_test.c

```diff
@@ -843,3 +843,320 @@ static int s_tls_client_channel_negotiation_success_fn(struct aws_allocator *all
+    (void)status;
+    struct shutdown_listener_tester *tester = arg;
+    /* destroy the listener */
+    aws_server_bootstrap_destroy_socket_listener(tester->server_bootstrap, tester->listener);
+    aws_mem_release(tester->outgoing_args->allocator, task);
+    AWS_FATAL_ASSERT(aws_socket_close(tester->socket) == AWS_OP_SUCCESS);
+}
+
+static void s_tester_client_connection_established_fool(struct aws_socket *socket, int error_code, void *user_data) {
+    /* connection is fooled~*/
+    (void)error_code;
+    struct shutdown_listener_tester *tester = user_data;
+    tester->socket = socket;
+
+    uint64_t run_at_ns;
+    aws_event_loop_current_clock_time(socket->event_loop, &run_at_ns);
+    run_at_ns += aws_timestamp_convert(1, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_NANOS, NULL);
+    struct aws_task *shutdown_listener_task =
+        aws_mem_acquire(tester->outgoing_args->allocator, sizeof(struct aws_task));
+    aws_task_init(shutdown_listener_task, s_shutdown_listener_task, tester, "wait_a_bit");
+    /* wait 1 sec for server side setup the channel then shut down the listener and close the socket */
+    aws_event_loop_schedule_task_future(socket->event_loop, shutdown_listener_task, run_at_ns);
+}
+
+static int s_tls_server_destroy_by_user_when_connection_is_in_processing_fn(
+    struct aws_allocator *allocator,
+    void *ctx) {
+    (void)ctx;
+
+    aws_tls_init_static_state(allocator);
+    struct aws_event_loop_group el_group;
+    ASSERT_SUCCESS(aws_event_loop_group_default_init(&el_group, allocator, 0));
+
+    struct aws_host_resolver resolver;
+    ASSERT_SUCCESS(aws_host_resolver_init_default(&resolver, allocator, 1, &el_group));
+
+    struct aws_mutex mutex = AWS_MUTEX_INIT;
+    struct aws_condition_variable condition_variable = AWS_CONDITION_VARIABLE_INIT;
+
+    struct aws_tls_ctx_options server_ctx_options;
+#ifdef __APPLE__
+    struct aws_byte_cursor pwd_cur = aws_byte_cursor_from_c_str("1234");
+    aws_tls_ctx_options_init_server_pkcs12_from_path(&server_ctx_options, allocator, "./unittests.p12", &pwd_cur);
+#else
+    aws_tls_ctx_options_init_default_server_from_path(
+        &server_ctx_options, allocator, "./unittests.crt", "./unittests.key");
+#endif /* __APPLE__ */
+    aws_tls_ctx_options_set_alpn_list(&server_ctx_options, "h2;http/1.1");
+
+    struct aws_tls_ctx *server_ctx = aws_tls_server_ctx_new(allocator, &server_ctx_options);
+    ASSERT_NOT_NULL(server_ctx);
+
+    struct tls_test_args incoming_args = {
+        .mutex = &mutex,
+        .allocator = allocator,
+        .condition_variable = &condition_variable,
+        .error_invoked = false,
+        .expects_error = false,
+        .rw_handler = NULL,
+        .server = true,
+        .tls_negotiated = false,
+        .shutdown_finished = false,
+    };
+
+    struct tls_test_args outgoing_args = {
+        .mutex = &mutex,
+        .allocator = allocator,
+        .condition_variable = &condition_variable,
```

Comment:
Man, it would be nice if the tests in this file didn't require sooooo much copy/paste. Like if there were a single s_tester_init() for all this common setup, like the tests in HTTP.

Don't do it in the PR, it would require changing too much unrelated code. But it would be a wise thing to do as a follow-up PR. So it's easier to add more tests in the future.

Reply:
...I also think about a tester_init() function will be much better.